### PR TITLE
FEATURE: Allow users to remove their vote

### PR DESCRIPTION
**Pull request by nbianca**
They can use the remove vote button or select the same option again for
single choice polls.

### CODE REVIEW

File: plugins/poll/assets/javascripts/widgets/discourse-poll.js.es6 [hunk 1]
```diff
@@ -636,24 +636,38 @@ createWidget("discourse-poll-buttons", {
...
+        showResultsButton = this.attach("button", {
+          className: "btn-default toggle-results",
+          label: "poll.show-results.label",
+          title: "poll.show-results.title",
+          icon: "far-eye",
+          action: "toggleResults",
+        });
+      }
+
+      contents.push(showResultsButton);
```
Comment (ZogStriP) [inline plugins/poll/assets/javascripts/widgets/discourse-poll.js.es6:658]: Do we need to `push` it if it's `undefined`?

File: plugins/poll/assets/javascripts/widgets/discourse-poll.js.es6 [hunk 2]
```diff
@@ -636,24 +636,38 @@ createWidget("discourse-poll-buttons", {
...
-            action: "toggleResults",
+            className: "btn-default remove-vote",
+            label: "poll.remove-vote.label",
+            title: "poll.remove-vote.title",
+            icon: "trash-alt",
+            action: "removeVote",
           })
         );
       }
+      contents.push(infoText);
```
Comment (ZogStriP) [inline plugins/poll/assets/javascripts/widgets/discourse-poll.js.es6:670]: Ditto ^^

File: plugins/poll/assets/javascripts/widgets/discourse-poll.js.es6 [hunk 3]
```diff
@@ -894,6 +908,28 @@ export default createWidget("discourse-poll", {
     this.state.showResults = !this.state.showResults;
   },
 
+  removeVote() {
+    const { attrs, state } = this;
+    state.loading = true;
+    return ajax("/polls/vote", {
+      type: "PUT",
```
Comment (ZogStriP) [inline plugins/poll/assets/javascripts/widgets/discourse-poll.js.es6:915]: Should probably use the `DELETE` verb and a different route on the server to handle the removal of a vote

File: plugins/poll/assets/javascripts/widgets/discourse-poll.js.es6 [hunk 4]
```diff
@@ -963,6 +999,10 @@ export default createWidget("discourse-poll", {
     }
 
     const { vote } = attrs;
+    if (!this.isMultiple() && vote.length === 1 && vote[0] === option.id) {
```
Comment (ZogStriP) [inline plugins/poll/assets/javascripts/widgets/discourse-poll.js.es6:1002]: Why couldn't they remove their vote on multiple choices polls?

File: plugins/poll/plugin.rb [hunk 5]
```diff
@@ -85,8 +85,6 @@ def vote(post_id, poll_name, options, user)
           available_options = poll.poll_options.map { |o| o.digest }.to_set
           options.select! { |o| available_options.include?(o) }
 
-          raise StandardError.new I18n.t("poll.requires_at_least_1_valid_option") if options.empty?
```
Comment (ZogStriP) [inline plugins/poll/plugin.rb:88]: I think we still need this since we should be using a different route for the vote removal

File: plugins/poll/assets/javascripts/widgets/discourse-poll.js.es6 [hunk 6]
```diff
@@ -636,24 +636,44 @@ createWidget("discourse-poll-buttons", {
...
+      if (attrs.hasSavedVote) {
         contents.push(
           this.attach("button", {
-            className: "btn-default toggle-results",
-            label: "poll.show-results.label",
-            title: "poll.show-results.title",
-            icon: "far-eye",
-            disabled: poll.voters === 0,
-            action: "toggleResults",
+            className: "btn-default remove-vote",
```
Comment (ZogStriP) [inline plugins/poll/assets/javascripts/widgets/discourse-poll.js.es6:665]: How does the button look like? Would it be better to make it red?

File: plugins/poll/plugin.rb [hunk 7]
```diff
@@ -398,6 +398,18 @@ def vote
       end
     end
 
+    def remove_vote
+      post_id   = params.require(:post_id)
+      poll_name = params.require(:poll_name)
+
+      begin
+        poll, options = DiscoursePoll::Poll.vote(post_id, poll_name, [], current_user, remove: true)
```
Comment (tgxworld) [inline plugins/poll/plugin.rb:406]: I'm personally not a huge fan of extending the `DiscoursePoll::Poll.vote` method to be able to remove vote because the original intend of the method is to cast a vote. A new method like `DiscoursePoll::Poll.remove_vote` would make it much clearer IMO.

File: plugins/poll/plugin.rb [hunk 8]
```diff
@@ -398,6 +398,18 @@ def vote
...
+        render json: { poll: poll, vote: options }
+      rescue StandardError => e
```
Comment (tgxworld) [inline plugins/poll/plugin.rb:408]: This is not related to your change but I wished we had use a custom error class instead. This is rescuing all errors whether they are related to poll vote or not.

**Review by tgxworld — APPROVED** [review]
@nbianca I really like the refactor that you did here. Things are properly organised and alot easier to follow.

File: plugins/poll/app/controllers/polls_controller.rb [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  before_action :ensure_logged_in, except: [:voters, :grouped_poll_results]
+
+  def vote
+    post_id = params.require(:post_id)
+    poll_name = params.require(:poll_name)
+    options = params.require(:options)
+
+    begin
+      poll, options = DiscoursePoll::Poll.vote(post_id, poll_name, options, current_user)
+      render json: { poll: poll, vote: options }
```
Comment (ZogStriP) [inline plugins/poll/app/controllers/polls_controller.rb:15]: Why is `vote` called `options`? 🤔

File: plugins/poll/app/controllers/polls_controller.rb [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+  before_action :ensure_logged_in, except: [:voters, :grouped_poll_results]
+
+  def vote
+    post_id = params.require(:post_id)
+    poll_name = params.require(:poll_name)
+    options = params.require(:options)
+
+    begin
+      poll, options = DiscoursePoll::Poll.vote(post_id, poll_name, options, current_user)
```
Comment (ZogStriP) [inline plugins/poll/app/controllers/polls_controller.rb:14]: Love that you've made a method for each operations 👏 

I would have put the `current_user` as the 1st parameter but that's nitpicking.

File: plugins/poll/app/controllers/polls_controller.rb [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+      render_json_error e.message
+    end
+  end
+
+  def voters
+    post_id = params.require(:post_id)
+    poll_name = params.require(:poll_name)
+    opts = params.permit(:limit, :page, :option_id)
+
+    post = Post.find_by(id: post_id)
```
Comment (ZogStriP) [inline plugins/poll/app/controllers/polls_controller.rb:51]: Since you're not using `post` anywhere else, there's no point in loading it using `find_by`. Maybe just use `exists?`

File: plugins/poll/spec/controllers/polls_controller_spec.rb [hunk 12]
```diff
@@ -382,51 +396,5 @@
 
       expect(json["voters"][first].size).to eq(1)
     end
-
-  end
-
-  describe '#current_user_voted' do
```
Comment (ZogStriP) [inline plugins/poll/spec/controllers/polls_controller_spec.rb:388]: Why was this removed?

File: plugins/poll/app/controllers/polls_controller.rb (hunk 9 above)
Comment (nbianca) [reply plugins/poll/app/controllers/polls_controller.rb:15]: I am not sure exactly why, but this is the name in JavaScript too. It makes sense because a vote is nothing more than a set of selected poll options.

File: plugins/poll/spec/controllers/polls_controller_spec.rb (hunk 12 above)
Comment (nbianca) [reply plugins/poll/spec/controllers/polls_controller_spec.rb:388]: It was unused.

**Review by ZogStriP — APPROVED** [review]
